File: src/High Performance Computing/PVL 1/py/jacobi_parallel.py
import numpy as np
import time
import argparse
from multiprocessing import Pool

def initialize_matrix(n):
    A = np.zeros((n, n))
    for i in range(n):
        for j in range(n):
            if i == j:
                A[i, j] = 20
            else:
                A[i, j] = -2 ** (-(2 * abs(i - j)))
    return A

def initialize_vector(n):
    return np.ones(n)

def compute_residual(A, x, b):
    return np.linalg.norm(b - np.dot(A, x))


def compute_row(i, A, b, x, n):
    sum_ = sum(A[i, j] * x[j] for j in range(n) if i != j)
    return (b[i] - sum_) / A[i, i]

def jacobi_parallel(A, b, tol, num_threads):
    n = len(b)
    x = np.zeros(n)
    x_new = np.zeros(n)
    residual = compute_residual(A, x, b)
    initial_residual = residual

    while residual > tol * initial_residual:
        with Pool(num_threads) as pool:
            x_new = np.array(pool.starmap(compute_row, [(i, A, b, x, n) for i in range(n)]))
        x = x_new.copy()
        residual = compute_residual(A, x, b)
    return x

if __name__ == "__main__":

    parser = argparse.ArgumentParser(description="Parallel Jacobi Method with Strong Scaling Test")
    parser.add_argument("-n", type=int, required=True, help="Size of the matrix (n x n)")
    parser.add_argument("-t", type=int, default=1, help="Number of threads")
    parser.add_argument("--test-scaling", action="store_true", help="Run strong scaling test")
    args = parser.parse_args()

    n = args.n
    num_threads = args.t
    tol = 1e-6

    A = initialize_matrix(n)
    b = initialize_vector(n)

    if args.test_scaling:
        for threads in [1, 2, 4, 8]:
            start_time = time.time()
            x_parallel = jacobi_parallel(A, b, tol, num_threads=threads)
            end_time = time.time()
            print(f"Threads: {threads}, Time: {end_time - start_time:.2f} seconds")
    else:

        start_time = time.time()
        x_parallel = jacobi_parallel(A, b, tol, num_threads=num_threads)
        end_time = time.time()
        print(f"Parallel Jacobi method completed for n={n}, threads={num_threads}.")
        print(f"Execution time: {end_time - start_time:.2f} seconds.")

File: src/High Performance Computing/PVL 1/cpp/jacobi_parallel.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <thread>
#include <mutex>
#include <numeric>
#include <chrono>
#include <functional>
#include <sstream>
#include <iomanip>

std::vector<std::vector<double>> initialize_matrix(int n) {
    std::vector<std::vector<double>> A(n, std::vector<double>(n, 0));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i == j) {
                A[i][j] = 20;
            } else {
                A[i][j] = -pow(2, -(2 * std::abs(i - j)));
            }
        }
    }
    return A;
}

std::vector<double> initialize_vector(int n) {
    return std::vector<double>(n, 1.0);
}

double compute_residual(const std::vector<std::vector<double>>& A, const std::vector<double>& x, const std::vector<double>& b) {
    int n = b.size();
    double norm = 0.0;
    for (int i = 0; i < n; ++i) {
        double sum = 0.0;
        for (int j = 0; j < n; ++j) {
            sum += A[i][j] * x[j];
        }
        norm += pow(b[i] - sum, 2);
    }
    return sqrt(norm);
}

double compute_row(int i, const std::vector<std::vector<double>>& A, const std::vector<double>& b, const std::vector<double>& x, int n) {
    double sum = 0.0;
    for (int j = 0; j < n; ++j) {
        if (i != j) {
            sum += A[i][j] * x[j];
        }
    }
    return (b[i] - sum) / A[i][i];
}

std::vector<double> jacobi_parallel(const std::vector<std::vector<double>>& A, const std::vector<double>& b, double tol, int num_threads) {
    int n = b.size();
    std::vector<double> x(n, 0.0);
    std::vector<double> x_new(n, 0.0);

    double residual = compute_residual(A, x, b);
    double initial_residual = residual;

    while (residual > tol * initial_residual) {
        std::vector<std::thread> threads;

        for (int t = 0; t < num_threads; ++t) {
            threads.emplace_back([&, t]() {
                for (int i = t; i < n; i += num_threads) {
                    x_new[i] = compute_row(i, A, b, x, n);
                }
            });
        }

        for (auto& th : threads) {
            th.join();
        }

        x = x_new;
        residual = compute_residual(A, x, b);
    }
    return x;
}

int main(int argc, char* argv[]) {
    int n = 0;
    int num_threads = 1;
    bool test_scaling = false;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-n" && i + 1 < argc) {
            n = std::stoi(argv[++i]);
        } else if (arg == "-t" && i + 1 < argc) {
            num_threads = std::stoi(argv[++i]);
        } else if (arg == "--test-scaling") {
            test_scaling = true;
        }
    }

    if (n <= 0) {
        std::cerr << "Matrix size (-n) must be specified and greater than 0.\n";
        return 1;
    }

    double tol = 1e-6;
    auto A = initialize_matrix(n);
    auto b = initialize_vector(n);

    if (test_scaling) {
        for (int threads : {1, 2, 4, 8}) {
            auto start = std::chrono::high_resolution_clock::now();
            jacobi_parallel(A, b, tol, threads);
            auto end = std::chrono::high_resolution_clock::now();
            std::chrono::duration<double> elapsed = end - start;
            std::cout << "Threads: " << threads << ", Time: " << std::fixed << std::setprecision(2) << elapsed.count() << " seconds\n";
        }
    } else {
        auto start = std::chrono::high_resolution_clock::now();
        auto result = jacobi_parallel(A, b, tol, num_threads);
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end - start;
        std::cout << "Parallel Jacobi method completed for n=" << n << ", threads=" << num_threads << ".\n";
        std::cout << "Execution time: " << std::fixed << std::setprecision(2) << elapsed.count() << " seconds\n";
    }

    return 0;
}
